### PR TITLE
Update graph operator overloading for subclasses

### PR DESCRIPTION
**Pull request by shreyasnagare**
This PR makes the set operations for rdflib.Graph subclasses (like `g1 = g1 + g2`) more intuitive and similar to the results of in-place set operations (like `g1 += g2`).

**Example**:

If you were to do a union of a `ConjunctiveGraph` (a subclass of Graph) instance with itself **or** any other `ConjunctiveGraph` instance using something like `g = g + g`, the instance would lose all the methods specific to the `ConjunctiveGraph`.

**Steps to reproduce the issue**:

```python
from rdflib import Graph

class MyGraph(Graph):
    def perform_reasoning(self):
        pass

g = MyGraph()
"perform_reasoning" in dir(g)
# True

g += g
"perform_reasoning" in dir(g)
# True

g = g + g
"perform_reasoning" in dir(g)
# False
```

## Proposed Changes
  - Use `type(self)()` instead of `Graph()` as the initial return value for set operations.
  - This would allow `Graph` to pass down similar behavior to its derived classes (unless overridden) without breaking any existing functionality.

### CODE REVIEW

**Review by nicholascar — APPROVED** [review]
Looks sensible to me. 

The only test failure is an address assignment, so not a real one.

**Comment by nicholascar** [comment]
@shreyasnagare thinks for this PR - looks good to me. Can you please add in your "steps to reproduce" as an actual test though? So test_graph_operator.py or similar? Thanks.

**Comment by nicholascar** [comment]
@shreyasnagare thanks for that rapid turn-around! Just waiting for another review from one of my co-maintainers and then we'll likely merge. 

Please do feel free to make more good updates like this!

**Review by aucampia — APPROVED** [review]

File: rdflib/graph.py
Comment (shreyasnagare) [reply rdflib/graph.py:562]: Oh, I do prefer using `type(self)` over `self.__class__`. I actually changed everything to  `self.__class__` after I saw it being used instead of `type(self)` in `term.py`. I was just trying to keep it consistent with rest of the code. I'll make the change right away. :)

**Comment by gtfierro** [comment]
Great to see this change, and thanks @shreyasnagare for finding and addressing it. I'd like to incorporate these changes in the `brickschema` package --- is there a timeline for the next release of RDFlib?

**Comment by nicholascar** [comment]
@gtfierro 6.0.0 is only a couple of days away! We merged in the JSON-LD plug-in, sorted out namespace issues and merged in heaps of smaller PRs like this. Even updated docco. Just waiting for some final CI/CD steps and we will have the package into PyPI. Again, only a few days!

**Comment by gtfierro** [comment]
Great news! Congrats on the release

**Pull request merged**
